ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java

```diff
@@ -17,96 +17,107 @@
-  private long tokens;
   private Long insertsSent;
   private long acksSent;
+  private boolean drainHandlerSet;
+  private Subscription subscription;
+  private boolean cancelled;
 
-  public AcksSubscriber(final HttpServerResponse response) {
-    this.response = Objects.requireNonNull(response);
+  public AcksSubscriber(final Context context, final HttpServerResponse response) {
+    super(context);
+    this.response = response;
   }
 
   @Override
-  public synchronized void onSubscribe(final Subscription subscription) {
-    Objects.requireNonNull(subscription);
-    if (this.subscription != null) {
-      throw new IllegalStateException("Already subscribed");
-    }
+  protected void afterSubscribe(final Subscription subscription) {
+    makeRequest(REQUEST_BATCH_SIZE);
     this.subscription = subscription;
-    checkRequestTokens();
   }
 
   @Override
-  public synchronized void onNext(final Void vo) {
-    if (tokens == 0) {
-      throw new IllegalStateException("Unsolicited data");
+  public void handleValue(final JsonObject value) {
+    checkContext();
+    if (cancelled) {
+      return;
     }
     response.write(ACK_RESPONSE_LINE);
     acksSent++;
-    tokens--;
     if (insertsSent != null && insertsSent == acksSent) {
       close();
     } else if (response.writeQueueFull()) {
-      response.drainHandler(v -> checkRequestTokens());
+      if (!drainHandlerSet) {
+        response.drainHandler(v -> {
+          drainHandlerSet = false;
+          checkMakeRequest();
+        });
+        drainHandlerSet = true;
+      }
     } else {
-      checkRequestTokens();
-    }
-  }
-
-  synchronized void insertsSent(final long num) {
-    this.insertsSent = num;
-    if (acksSent == num) {
-      close();
+      checkMakeRequest();
     }
   }
 
-  private void close() {
+  @Override
+  public void handleComplete() {
     response.end();
-    subscription.cancel();
-  }
-
-  private void checkRequestTokens() {
-    if (tokens == 0) {
-      tokens = BATCH_SIZE;
-      subscription.request(BATCH_SIZE);
-    }
   }
 
   @Override
-  public synchronized void onError(final Throwable t) {
+  public void handleError(final Throwable t) {
+    if (cancelled) {
+      return;
+    }
     log.error("Error in processing inserts", t);
     final JsonObject err = new JsonObject().put("status", "error")
         .put("errorCode", ERROR_CODE_INTERNAL_ERROR)
         .put("message", "Error in processing inserts");
-    subscription.cancel();
     response.end(err.toBuffer());
   }
 
-  @Override
-  public synchronized void onComplete() {
+  public void cancel() {
+    checkContext();
+    cancelled = true;
+    if (subscription != null) {
+      subscription.cancel();
+    }
+  }
+
+  private void checkMakeRequest() {
+    if (acksSent % REQUEST_BATCH_SIZE == 0) {
+      makeRequest(REQUEST_BATCH_SIZE);
+    }
+  }
+
+  private void close() {
     response.end();
+    complete();
+  }
+
+  void insertsSent(final long num) {
```

Comment:
same comment as the previous PR, I'm not sure this is out of date. **tl;dr** I think we should rename this `requestClose(final long expectedAcks)` and to change `insertsSent` to `expectedAcks`

---

It took me some time to understand what this API does. Does it indicate that we've completed all the inserts we want to send? Do we expect it to be called more than once? Maybe we can name it `requestClose(final long expectedAcks)`. Would also help to have javadoc around here.

Also instead of using a single `Long` to indicate two things: both that we should close and the value of acks expected, I think it would be safer (avoid unboxing a null value) and easier to read if we split it into a flag boolean `closeRequested` and long expectedAcks.